components/serviceMap/customNodes.tsx

```diff
@@ -0,0 +1,441 @@
+import { Handle, MarkerType, Position } from "reactflow";
+import IconGripVertical from "tabler-icons/tsx/grip-vertical.tsx";
+import IconDatabase from "tabler-icons/tsx/database.tsx";
+import "flowbite";
+import "twind";
+import { useState } from "preact/hooks";
+import { Audience, OperationType } from "snitch-protos/protos/common.ts";
+import { NodeMenu, ServiceNodeMenu } from "./nodeMenu.tsx";
+import { NodeData } from "../../islands/serviceMap.tsx";
+import { ProducerIcon } from "../icons/producer.tsx";
+import { ConsumerIcon } from "../icons/consumer.tsx";
+import { titleCase } from "../../lib/utils.ts";
+import { Pipeline } from "snitch-protos/protos/pipeline.ts";
+import { PipelineInfo } from "snitch-protos/protos/info.ts";
+
+export type AudiencePipeline = Audience & { pipeline?: Pipeline };
+
+export type NodeData = {
+  label: string;
+  audience: AudiencePipeline;
+  groupCount?: number;
+};
+
+export type FlowNode = {
+  id: string;
+  type?: string;
+  dragHandle: string;
+  position?: {
+    x: number;
+    y: number;
+  };
+  sourcePosition?: string;
+  targetPosition?: string;
+  data: NodeData;
+  parentNode?: string;
+  extent?: string;
+  style?: any;
+};
+
+export type GroupCount = {
+  producer: number;
+  consumer: number;
+};
+//
+// group counts per service are used for vertical offset layout positions
+export type NodesMap = {
+  nodes: Map<string, FlowNode>;
+  groupCount: Map<string, GroupCount>;
+};
+
+export type FlowEdge = {
+  id: string;
+  source: string;
+  target: string;
+  markerEnd: any;
+  style: any;
+};
+
+export const xOffset = (serviceCount: number) => serviceCount > 1 ? 800 : 0;
+
+export const mapOperation = (
```

Comment:
The mappings below are all fairly gnarly. The executive summary is that we are mapping our array of audience structures to flow service/operation/component nodes. While we are at it we have to keep counts of services and operation groups per service so we can calculate x and y axis offsets dynamically.